### PR TITLE
xwm: add loop detection for read_surface_parent

### PR DESCRIPTION
**Pull request by Emantor**
Implement a simple loop detection while trying to retrieve the parent
for a TRANSIENT_FOR window.

Fixes swaywm/sway#4624

### CODE REVIEW

File: xwayland/xwm.c [hunk 1]
```diff
@@ -467,20 +467,39 @@ static void read_surface_title(struct wlr_xwm *xwm,
 	wlr_signal_emit_safe(&xsurface->events.set_title, xsurface);
 }
 
+static bool check_transient_loop(struct wlr_xwayland_surface *candidate,
```
Comment (emersion) [inline xwayland/xwm.c:470]: This could be renamed `has_parent`

File: xwayland/xwm.c [hunk 2]
```diff
@@ -467,20 +467,39 @@ static void read_surface_title(struct wlr_xwm *xwm,
...
+		struct wlr_xwayland_surface *parent) {
+	while (parent) {
+		if (candidate == parent)
```
Comment (emersion) [inline xwayland/xwm.c:473]: Style: braces are mandatory

File: xwayland/xwm.c [hunk 3]
```diff
@@ -467,20 +467,39 @@ static void read_surface_title(struct wlr_xwm *xwm,
...
 
 	xcb_window_t *xid = xcb_get_property_value(reply);
 	if (xid != NULL) {
-		xsurface->parent = lookup_surface(xwm, *xid);
+		candidate = lookup_surface(xwm, *xid);
 	} else {
 		xsurface->parent = NULL;
 	}
 
+	if (!check_transient_loop(xsurface, candidate)) {
```
Comment (emersion) [inline xwayland/xwm.c:497]: `check_transient_loop` is called even if `xsurface->parent` is set to NULL above, which is surprising. Can we move this in the `if (xid != NULL)` block?

File: xwayland/xwm.c (hunk 1 above)
Comment (Emantor) [reply xwayland/xwm.c:470]: I don't think `has_parent` fits here, maybe `is_parent`?

File: xwayland/xwm.c (hunk 3 above)
Comment (Emantor) [reply xwayland/xwm.c:497]: Yep, definitely.

File: xwayland/xwm.c (hunk 1 above)
Comment (emersion) [reply xwayland/xwm.c:470]: `has_parent` would be fine if the args were reversed. BTW maybe we can find better names for these args.

Comment (Emantor) [reply xwayland/xwm.c]: I'll go with `parent` and `child` and `has_parent`. Also renamed the `candidate` variable to `found_parent`.

**Comment by crocket** [comment]
Check https://github.com/swaywm/sway/issues/4624#issuecomment-695192840

**Comment by Emantor** [comment]
Gentle Ping.

**Review by emersion — APPROVED** [review]
Thanks!

**Pull request merged**